File: apps/mobile/lib/features/mess/domain/mess.dart
import '../../nutrition/domain/entities/food.dart';
import '../../nutrition/domain/entities/food_log.dart';

/// Wire shapes of `@fitos/contracts` mess.ts (Phase 9). The server mirrors
/// MessIT and parses it; the app only ever sees these normalised shapes,
/// never MessIT's own format. Hand-written (ADR-004): the menu resolution
/// is a union on `kind`.

enum DietClass {
  veg('veg', 'Veg'),
  egg('egg', 'Egg'),
  nonveg('nonveg', 'Non-veg'),

  /// Deliberate: the classifier could not tell. Never shown as veg.
  unknown('unknown', 'Diet not known');

  const DietClass(this.wire, this.label);
  final String wire;
  final String label;

  static DietClass fromWire(String? w) =>
      values.firstWhere((d) => d.wire == w, orElse: () => DietClass.unknown);
}

enum MirrorError {
  unreachable('unreachable'),
  malformed('malformed');

  const MirrorError(this.wire);
  final String wire;

  static MirrorError? fromWire(String? w) =>
      w == null ? null : values.where((e) => e.wire == w).firstOrNull;
}

/// How fresh the SERVER's copy of one endpoint is — MessIT has no timestamp
/// of its own. `stale`: no successful fetch in 24 h.
class MessFreshness {
  const MessFreshness({
    required this.lastSuccessAt,
    required this.lastAttemptAt,
    required this.lastError,
    required this.stale,
    required this.latestPublishedDate,
  });

  factory MessFreshness.fromJson(Map<String, dynamic> json) => MessFreshness(
        lastSuccessAt: json['lastSuccessAt'] as String?,
        lastAttemptAt: json['lastAttemptAt'] as String?,
        lastError: MirrorError.fromWire(json['lastError'] as String?),
        stale: json['stale'] as bool,
        latestPublishedDate: json['latestPublishedDate'] as String?,
      );

  final String? lastSuccessAt;
  final String? lastAttemptAt;
  final MirrorError? lastError;
  final bool stale;
  final String? latestPublishedDate;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'lastSuccessAt': lastSuccessAt,
        'lastAttemptAt': lastAttemptAt,
        'lastError': lastError?.wire,
        'stale': stale,
        'latestPublishedDate': latestPublishedDate,
      };
}

class Mess {
  const Mess({
    required this.code,
    required this.providerSlug,
    required this.hostelId,
    required this.hostelLabel,
    required this.messId,
    required this.messLabel,
    required this.servesNonVeg,
    required this.freshness,
  });

  factory Mess.fromJson(Map<String, dynamic> json) => Mess(
        code: json['code'] as String,
        providerSlug: json['providerSlug'] as String,
        hostelId: json['hostelId'] as String,
        hostelLabel: json['hostelLabel'] as String,
        messId: json['messId'] as String,
        messLabel: json['messLabel'] as String,
        servesNonVeg: json['servesNonVeg'] as bool,
        freshness:
            MessFreshness.fromJson(json['freshness'] as Map<String, dynamic>),
      );

  final String code;
  final String providerSlug;
  final String hostelId;
  final String hostelLabel;
  final String messId;
  final String messLabel;
  final bool servesNonVeg;
  final MessFreshness freshness;

  /// "Men's Hostel · Vegetarian".
  String get label => '$hostelLabel · $messLabel';

  Map<String, dynamic> toJson() => <String, dynamic>{
        'code': code,
        'providerSlug': providerSlug,
        'hostelId': hostelId,
        'hostelLabel': hostelLabel,
        'messId': messId,
        'messLabel': messLabel,
        'servesNonVeg': servesNonVeg,
        'freshness': freshness.toJson(),
      };
}

class MessProviderInfo {
  const MessProviderInfo({
    required this.slug,
    required this.displayName,
    required this.status,
  });

  factory MessProviderInfo.fromJson(Map<String, dynamic> json) =>
      MessProviderInfo(
        slug: json['slug'] as String,
        displayName: json['displayName'] as String,
        status: json['status'] as String,
      );

  final String slug;
  final String displayName;
  final String status;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'slug': slug,
        'displayName': displayName,
        'status': status,
      };
}

class MessesResponse {
  const MessesResponse({required this.provider, required this.items});

  factory MessesResponse.fromJson(Map<String, dynamic> json) => MessesResponse(
        provider:
            MessProviderInfo.fromJson(json['provider'] as Map<String, dynamic>),
        items: (json['items'] as List<dynamic>)
            .map((e) => Mess.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  final MessProviderInfo provider;
  final List<Mess> items;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'provider': provider.toJson(),
        'items': items.map((m) => m.toJson()).toList(),
      };
}

/// How the menu on screen was found (§14.4) — always shown, never hidden.
sealed class MenuResolution {
  const MenuResolution(this.date);

  factory MenuResolution.fromJson(Map<String, dynamic> json) =>
      switch (json['kind']) {
        'exact' => ExactMenu(json['date'] as String),
        'cycle-inferred' => InferredMenu(
            json['date'] as String,
            sourceDate: json['sourceDate'] as String,
            cycleLengthDays: json['cycleLengthDays'] as int,
          ),
        _ => UnavailableMenu(
            json['date'] as String,
            latestAvailable: json['latestAvailable'] as String?,
          ),
      };

  final String date;

  Map<String, dynamic> toJson();
}

/// The mess published this date.
class ExactMenu extends MenuResolution {
  const ExactMenu(super.date);

  @override
  Map<String, dynamic> toJson() =>
      <String, dynamic>{'kind': 'exact', 'date': date};
}

/// Not published: the menu repeats every [cycleLengthDays], so this is the
/// menu of [sourceDate]. Never presented as the published menu.
class InferredMenu extends MenuResolution {
  const InferredMenu(
    super.date, {
    required this.sourceDate,
    required this.cycleLengthDays,
  });

  final String sourceDate;
  final int cycleLengthDays;

  @override
  Map<String, dynamic> toJson() => <String, dynamic>{
        'kind': 'cycle-inferred',
        'date': date,
        'sourceDate': sourceDate,
        'cycleLengthDays': cycleLengthDays,
      };
}

/// Neither published nor inferable.
class UnavailableMenu extends MenuResolution {
  const UnavailableMenu(super.date, {required this.latestAvailable});

  final String? latestAvailable;

  @override
  Map<String, dynamic> toJson() => <String, dynamic>{
        'kind': 'unavailable',
        'date': date,
        'latestAvailable': latestAvailable,
      };
}

class MessDish {
  const MessDish({
    required this.slug,
    required this.name,
    required this.label,
    required this.diet,
    required this.role,
    required this.alternatives,
    required this.isAmbient,
    required this.nutrition,
    required this.correctionPending,
  });

  factory MessDish.fromJson(Map<String, dynamic> json) => MessDish(
        slug: json['slug'] as String,
        name: json['name'] as String,
        label: json['label'] as String?,
        diet: DietClass.fromWire(json['diet'] as String?),
        role: json['role'] as String,
        alternatives: (json['alternatives'] as List<dynamic>).cast<String>(),
        isAmbient: json['isAmbient'] as bool,
        nutrition: json['nutrition'] == null
            ? null
            : messRowFromJson(json['nutrition'] as Map<String, dynamic>),
        correctionPending: json['correctionPending'] as bool,
      );

  final String slug;
  final String name;
  final String? label;
  final DietClass diet;
  final String role;
  final List<String> alternatives;

  /// Bread, tea, jam… at every meal: loggable, shown quietly.
  final bool isAmbient;

  /// The estimate per serving as a food row; null = no estimate (not
  /// tap-loggable).
  final FoodNutrition? nutrition;

  /// You reported this dish; the report is awaiting review (owner D17).
  final bool correctionPending;

  /// The dish as a food, so the Phase 8 portion step can take it. It is
  /// never stored as a food: logging sends the dish slug.
  Food asFood() => Food(
        id: 'mess:$slug',
        slug: slug,
        name: name,
        brand: null,
        barcode: null,
        source: FoodSource.estimated,
        sourceRef: null,
        isVerified: false,
        isCustom: false,
        aliases: const <String>[],
        nutrition: nutrition == null ? const [] : [nutrition!],
      );

  Map<String, dynamic> toJson() => <String, dynamic>{
        'slug': slug,
        'name': name,
        'label': label,
        'diet': diet.wire,
        'role': role,
        'alternatives': alternatives,
        'isAmbient': isAmbient,
        'nutrition': nutrition == null ? null : messRowToJson(nutrition!),
        'correctionPending': correctionPending,
      };
}

class MessMeal {
  const MessMeal({
    required this.slot,
    required this.rawMenu,
    required this.dishes,
  });

  factory MessMeal.fromJson(Map<String, dynamic> json) => MessMeal(
        slot: MealSlot.fromWire(json['slot'] as String),
        rawMenu: json['rawMenu'] as String,
        dishes: (json['dishes'] as List<dynamic>)
            .map((e) => MessDish.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  final MealSlot slot;

  /// Exactly what the mess published for this meal (owner D20).
  final String rawMenu;
  final List<MessDish> dishes;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'slot': slot.wire,
        'rawMenu': rawMenu,
        'dishes': dishes.map((d) => d.toJson()).toList(),
      };
}

class MessLoggedDish {
  const MessLoggedDish({
    required this.dishSlug,
    required this.mealSlot,
    required this.clientLogId,
  });

  factory MessLoggedDish.fromJson(Map<String, dynamic> json) => MessLoggedDish(
        dishSlug: json['dishSlug'] as String,
        mealSlot: MealSlot.fromWire(json['mealSlot'] as String),
        clientLogId: json['clientLogId'] as String,
      );

  final String dishSlug;
  final MealSlot mealSlot;
  final String clientLogId;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'dishSlug': dishSlug,
        'mealSlot': mealSlot.wire,
        'clientLogId': clientLogId,
      };
}

class MessMenu {
  const MessMenu({
    required this.mess,
    required this.date,
    required this.today,
    required this.resolution,
    required this.meals,
    required this.logged,
  });

  factory MessMenu.fromJson(Map<String, dynamic> json) => MessMenu(
        mess: Mess.fromJson(json['mess'] as Map<String, dynamic>),
        date: json['date'] as String,
        today: json['today'] as String,
        resolution: MenuResolution.fromJson(
          json['resolution'] as Map<String, dynamic>,
        ),
        meals: (json['meals'] as List<dynamic>)
            .map((e) => MessMeal.fromJson(e as Map<String, dynamic>))
            .toList(),
        logged: (json['logged'] as List<dynamic>)
            .map((e) => MessLoggedDish.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  final Mess mess;
  final String date;
  final String today;
  final MenuResolution resolution;
  final List<MessMeal> meals;
  final List<MessLoggedDish> logged;

  MessMeal? meal(MealSlot slot) =>
      meals.where((m) => m.slot == slot).firstOrNull;

  bool isLogged(String slug) => logged.any((l) => l.dishSlug == slug);

  Map<String, dynamic> toJson() => <String, dynamic>{
        'mess': mess.toJson(),
        'date': date,
        'today': today,
        'resolution': resolution.toJson(),
        'meals': meals.map((m) => m.toJson()).toList(),
        'logged': logged.map((l) => l.toJson()).toList(),
      };
}

/* --------------------------------------------------------- corrections -- */

enum CorrectionField {
  kcal('kcal', 'Calories', 'kcal'),
  protein('protein', 'Protein', 'g'),
  carb('carb', 'Carbs', 'g'),
  fat('fat', 'Fat', 'g'),
  diet('diet', 'Diet type', ''),
  other('other', 'Something else', '');

  const CorrectionField(this.wire, this.label, this.unit);
  final String wire;
  final String label;
  final String unit;

  bool get isMacro => unit.isNotEmpty;
}

/// A report that a dish's estimate is wrong (owner D17): stored as pending,
/// changes nothing until a later review.
class MessCorrectionRequest {
  const MessCorrectionRequest({
    required this.clientCorrectionId,
    required this.field,
    this.low,
    this.high,
    this.diet,
    this.note,
  });

  final String clientCorrectionId;
  final CorrectionField field;
  final double? low;
  final double? high;
  final DietClass? diet;
  final String? note;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'clientCorrectionId': clientCorrectionId,
        'field': field.wire,
        if (low != null) 'low': low,
        if (high != null) 'high': high,
        if (diet != null) 'diet': diet!.wire,
        if (note != null) 'note': note,
      };
}

class MessCorrection {
  const MessCorrection({
    required this.id,
    required this.clientCorrectionId,
    required this.dishSlug,
    required this.field,
    required this.low,
    required this.high,
    required this.diet,
    required this.note,
    required this.status,
    required this.createdAt,
  });

  factory MessCorrection.fromJson(Map<String, dynamic> json) => MessCorrection(
        id: json['id'] as String,
        clientCorrectionId: json['clientCorrectionId'] as String,
        dishSlug: json['dishSlug'] as String,
        field: CorrectionField.values.firstWhere(
          (f) => f.wire == json['field'],
          orElse: () => CorrectionField.other,
        ),
        low: (json['low'] as num?)?.toDouble(),
        high: (json['high'] as num?)?.toDouble(),
        diet: json['diet'] == null
            ? null
            : DietClass.fromWire(json['diet'] as String),
        note: json['note'] as String?,
        status: json['status'] as String,
        createdAt: json['createdAt'] as String,
      );

  final String id;
  final String clientCorrectionId;
  final String dishSlug;
  final CorrectionField field;
  final double? low;
  final double? high;
  final DietClass? diet;
  final String? note;
  final String status;
  final String createdAt;

  Map<String, dynamic> toJson() => <String, dynamic>{
        'id': id,
        'clientCorrectionId': clientCorrectionId,
        'dishSlug': dishSlug,
        'field': field.wire,
        'low': low,
        'high': high,
        'diet': diet?.wire,
        'note': note,
        'status': status,
        'createdAt': createdAt,
      };
}

File: packages/contracts/src/mess.ts
/**
 * VIT mess contracts (Phase 9, §14/§31). The app never sees MessIT's wire
 * format: the server mirrors it, parses it with packages/core, and serves
 * these normalised shapes. Every menu says how it was resolved (published,
 * inferred from the cycle, or unavailable) and how fresh the server's copy is.
 */
import { z } from 'zod';

import { dishSlugSchema, messCodeSchema, messDishNutritionSchema } from './mess-common.js';
import { mealSlotSchema } from './nutrition-log.js';
import { isoDateSchema } from './profile.js';

export * from './mess-common.js';

/** packages/core `DietClass`. `unknown` is deliberate: never shown as veg. */
export const DIET_CLASSES = ['veg', 'egg', 'nonveg', 'unknown'] as const;
export const dietClassSchema = z.enum(DIET_CLASSES);
export type DietClass = z.infer<typeof dietClassSchema>;

/** packages/core `DishRole`. */
export const DISH_ROLES = [
  'staple', 'protein', 'legume', 'vegetable', 'dairy', 'fruit', 'sweet', 'fried', 'beverage', 'condiment', 'other',
] as const;
export const dishRoleSchema = z.enum(DISH_ROLES);

export const MESS_PROVIDER_STATUSES = ['active', 'disabled'] as const;
export const messProviderSchema = z.object({
  slug: z.string().min(1),
  displayName: z.string().min(1),
  status: z.enum(MESS_PROVIDER_STATUSES),
});
export type MessProviderInfo = z.infer<typeof messProviderSchema>;

export const messProvidersResponseSchema = z.object({ items: z.array(messProviderSchema) });

export const MIRROR_ERRORS = ['unreachable', 'malformed'] as const;

/**
 * How fresh the server's copy of one endpoint is (owner D4/D19). MessIT has
 * no timestamp of its own; this is when WE last fetched it. `stale` = no
 * successful fetch in 24 h (or ever). `latestPublishedDate` is the last date
 * the stored copies publish.
 */
export const messFreshnessSchema = z.object({
  lastSuccessAt: z.string().datetime().nullable(),
  lastAttemptAt: z.string().datetime().nullable(),
  lastError: z.enum(MIRROR_ERRORS).nullable(),
  stale: z.boolean(),
  latestPublishedDate: isoDateSchema.nullable(),
});
export type MessFreshness = z.infer<typeof messFreshnessSchema>;

export const messSchema = z.object({
  code: messCodeSchema,
  providerSlug: z.string().min(1),
  hostelId: z.string().min(1),
  hostelLabel: z.string().min(1),
  messId: z.string().min(1),
  messLabel: z.string().min(1),
  servesNonVeg: z.boolean(),
  freshness: messFreshnessSchema,
});
export type Mess = z.infer<typeof messSchema>;

export const messesResponseSchema = z.object({
  provider: messProviderSchema,
  items: z.array(messSchema),
});
export type MessesResponse = z.infer<typeof messesResponseSchema>;

export const providerSlugParamsSchema = z.object({ slug: z.string().min(1).max(40) }).strict();

/** packages/core `MenuResolution` — surfaced, never hidden (§14.4). */
export const menuResolutionSchema = z.discriminatedUnion('kind', [
  z.object({ kind: z.literal('exact'), date: isoDateSchema }),
  z.object({
    kind: z.literal('cycle-inferred'),
    date: isoDateSchema,
    sourceDate: isoDateSchema,
    cycleLengthDays: z.number().int().positive(),
  }),
  z.object({ kind: z.literal('unavailable'), date: isoDateSchema, latestAvailable: isoDateSchema.nullable() }),
]);
export type MenuResolution = z.infer<typeof menuResolutionSchema>;

export const messDishSchema = z.object({
  slug: dishSlugSchema,
  name: z.string().min(1),
  /** The mess's own label, e.g. "Non Veg"; null when unlabelled. */
  label: z.string().nullable(),
  diet: dietClassSchema,
  role: dishRoleSchema,
  alternatives: z.array(z.string()),
  /** Bread, tea, jam… present at every meal: loggable, shown quietly. */
  isAmbient: z.boolean(),
  /** Null when there is no estimate for this dish: then it cannot be tap-logged. */
  nutrition: messDishNutritionSchema.nullable(),
  /** You reported this dish's nutrition and the report is awaiting review (owner D17). */
  correctionPending: z.boolean(),
});
export type MessDish = z.infer<typeof messDishSchema>;

export const messMealSchema = z.object({
  slot: mealSlotSchema,
  /** Exactly what the mess published for this meal (owner D20). */
  rawMenu: z.string(),
  dishes: z.array(messDishSchema),
});
export type MessMeal = z.infer<typeof messMealSchema>;

export const messMenuQuerySchema = z
  .object({
    /** yyyy-mm-dd; defaults to today in your zone. */
    date: isoDateSchema.optional(),
    /** A mess code; defaults to your configured mess (owner D14). */
    mess: messCodeSchema.optional(),
  })
  .strict();
export type MessMenuQuery = z.infer<typeof messMenuQuerySchema>;

/** A mess dish you logged on this day (for the "logged" mark). */
export const messLoggedDishSchema = z.object({
  dishSlug: dishSlugSchema,
  mealSlot: mealSlotSchema,
  clientLogId: z.string().uuid(),
});

export const messMenuSchema = z.object({
  mess: messSchema,
  date: isoDateSchema,
  /** Today in your zone. */
  today: isoDateSchema,
  resolution: menuResolutionSchema,
  meals: z.array(messMealSchema),
  logged: z.array(messLoggedDishSchema),
});
export type MessMenu = z.infer<typeof messMenuSchema>;

/* --------------------------------------------------------- corrections -- */

export const MESS_CORRECTION_FIELDS = ['kcal', 'protein', 'carb', 'fat', 'diet', 'other'] as const;
export const messCorrectionFieldSchema = z.enum(MESS_CORRECTION_FIELDS);
export const MESS_CORRECTION_STATUSES = ['pending', 'accepted', 'rejected'] as const;

export const dishSlugParamsSchema = z.object({ slug: dishSlugSchema }).strict();

/**
 * A report that a dish's estimate is wrong (owner D17). Stored as pending; it
 * changes nothing until a later review accepts it. Macro fields give the
 * range you believe (per serving); `diet` gives the right class; `other`
 * needs a note. Retry-safe by `clientCorrectionId`.
 */
export const messCorrectionRequestSchema = z
  .object({
    clientCorrectionId: z.string().uuid(),
    field: messCorrectionFieldSchema,
    low: z.number().finite().nonnegative().max(5000).optional(),
    high: z.number().finite().nonnegative().max(5000).optional(),
    diet: z.enum(['veg', 'egg', 'nonveg']).optional(),
    note: z.string().trim().min(1).max(280).optional(),
  })
  .strict()
  .superRefine((r, ctx) => {
    const macro = r.field === 'kcal' || r.field === 'protein' || r.field === 'carb' || r.field === 'fat';
    if (macro) {
      if (r.low === undefined || r.high === undefined) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['low'], message: 'give the low and high you believe' });
      } else if (r.low > r.high) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['high'], message: 'high must not be below low' });
      }
      if (r.diet !== undefined) ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['diet'], message: 'only for field diet' });
    } else {
      if (r.low !== undefined || r.high !== undefined) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['low'], message: 'only for kcal, protein, carb or fat' });
      }
      if (r.field === 'diet' && r.diet === undefined) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['diet'], message: 'give the right diet class' });
      }
      if (r.field === 'other' && r.note === undefined) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['note'], message: 'say what is wrong' });
      }
      if (r.field === 'other' && r.diet !== undefined) {
        ctx.addIssue({ code: z.ZodIssueCode.custom, path: ['diet'], message: 'only for field diet' });
      }
    }
  });
export type MessCorrectionRequest = z.infer<typeof messCorrectionRequestSchema>;

export const messCorrectionSchema = z.object({
  id: z.string().uuid(),
  clientCorrectionId: z.string().uuid(),
  dishSlug: dishSlugSchema,
  field: messCorrectionFieldSchema,
  low: z.number().nullable(),
  high: z.number().nullable(),
  diet: z.enum(['veg', 'egg', 'nonveg']).nullable(),
  note: z.string().nullable(),
  status: z.enum(MESS_CORRECTION_STATUSES),
  createdAt: z.string().datetime(),
});
export type MessCorrection = z.infer<typeof messCorrectionSchema>;
